### PR TITLE
TYP: datetimelike

### PR DESCRIPTION
**Pull request by jbrockmendel**
use np.putmask in a few places, as it is appreciably more performant than `__setitem__`

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -755,28 +760,30 @@ def isin(self, values) -> np.ndarray:
     # ------------------------------------------------------------------
     # Null Handling
 
-    def isna(self):
+    def isna(self) -> np.ndarray:
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:763]: we dont' have the typing ability to do `np.ndarray[bool]` yet right?

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -996,7 +1006,7 @@ def _add_timedeltalike_scalar(self, other):
         if isna(other):
             # i.e np.timedelta64("NaT"), not recognized by delta_to_nanoseconds
             new_values = np.empty(self.shape, dtype="i8")
-            new_values[:] = iNaT
+            new_values.fill(iNaT)
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:1009]: can use np.full here

Comment (jreback) [reply pandas/core/arrays/datetimelike.py:1009]: followon ok too (as many places where this is done), though prob easy to do here

**Comment by jreback** [comment]
might actually be some perf boosts here due to using putmask

**Comment by jbrockmendel** [comment]
> might actually be some perf boosts here due to using putmask

yep thats the idea

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:763]: not yet, correct

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:1009]: yah, we use .fill way more often, to the point that i sometimes forget that np.full is a thing

**Comment by jreback** [comment]
thanks

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/datetimelike.py:763]: looks like will be in numpy 1.21 https://github.com/numpy/numpy/pull/17719

**Pull request merged**